src/test/java/uk/gov/hmcts/reform/sendletter/tasks/SerialTaskRunnerLockingTest.java

```diff
@@ -0,0 +1,153 @@
+public class SerialTaskRunnerLockingTest {
+
+    @Mock
+    private DataSource source;
+    @Mock
+    private Connection connection;
+    @Mock
+    private Statement statement;
+    private SerialTaskRunner taskRunner;
+
+    @Before
+    public void setup() throws SQLException {
+        when(connection.createStatement()).thenReturn(statement);
+        when(source.getConnection()).thenReturn(connection);
+
+        taskRunner = SerialTaskRunner.get(source);
+    }
+
+    @Test
+    public void successfully_locks_and_unlocks() throws SQLException {
+        lockedWithSuccess();
+        unlockedWithSuccess();
+
+        taskRunner.tryRun(1, () -> {
+        });
+    }
+
+    @Test
+    public void successfully_locks_and_fails_to_unlock() throws SQLException {
+        lockedWithSuccess();
+        unlockedWithFailure();
+
+        taskRunner.tryRun(1, () -> {
+        });
+    }
+
+    @Test
+    public void fails_to_lock_and_unlock() throws SQLException {
+        lockedWithFailure();
+        unlockedWithFailure();
+
+        taskRunner.tryRun(1, () -> {
+        });
+    }
+
+    @Test
+    public void failure_to_release_lock() throws SQLException {
+        lockedWithFailure();
+        unlockedWithSuccess();
+
+        taskRunner.tryRun(1, () -> {
+        });
+    }
+
+    @SuppressWarnings("unchecked")
+    @Test(expected = SQLException.class)
+    public void datasource_throws_connection_error() throws SQLException {
+        when(source.getConnection()).thenThrow(SQLException.class);
+
+        taskRunner.tryRun(1, () -> {
+        });
+    }
+
+    @SuppressWarnings("unchecked")
+    @Test(expected = SQLException.class)
+    public void locking_throws_exception_successfully_unlocks() throws SQLException {
+        when(statement.executeQuery(startsWith("SELECT pg_try_advisory_lock"))).thenThrow(SQLException.class);
+
+        taskRunner.tryRun(1, () -> {
+        });
+    }
+
+    @SuppressWarnings("unchecked")
+    @Test(expected = RuntimeException.class)
+    public void task_throws_exception_successfully_unlocks() throws SQLException {
+        lockedWithSuccess();
+        unlockedWithSuccess();
+
+        taskRunner.tryRun(1, () -> {
+            throw new RuntimeException();
+        });
+    }
+
+    @SuppressWarnings("unchecked")
+    @Test(expected = RuntimeException.class)
+    public void task_throws_exception_fails_to_unlock() throws SQLException {
+        lockedWithSuccess();
+        unlockedWithFailure();
+
+        taskRunner.tryRun(1, () -> {
+            throw new RuntimeException();
+        });
+    }
+
+    @SuppressWarnings("unchecked")
+    @Test(expected = SQLException.class)
+    public void executing_query_throws_exception() throws SQLException {
+        when(connection.createStatement()).thenThrow(SQLException.class);
+        unlockedWithFailure();
+
+        taskRunner.tryRun(1, () -> {
+        });
+    }
+
+    private void unlockedWithSuccess() throws SQLException {
```

Comment:
All methods starting from this one are responsible for setting up mocks. It would be best if their names said that, e.g. `setupSuccessfulUnlocking` or sth.